dagshub/mlflow/patch.py

```diff
@@ -0,0 +1,176 @@
+import inspect
+import logging
+import sys
+import traceback
+from types import ModuleType, FunctionType
+from typing import Callable, Optional, Dict, Set
+
+from tenacity import Retrying, stop_after_delay
+
+logger = logging.getLogger(__name__)
+
+
+class MlflowMonkeyPatch:
+    context_manager_functions = ["start_run"]
+    classes_to_patch = ["mlflow.tracking.client.MlflowClient"]
+
+    def __init__(self):
+        self.original_func_lookup: Dict[str, Callable] = {}
+        self.patched_modules: Set[str] = set()
+
+    def _wrap_func(self, fn: Callable) -> Callable:
+        from mlflow.exceptions import MlflowException
+
+        def is_top_level_mlflow_call() -> bool:
+            tb = traceback.extract_stack()
+
+            # Count the times `wrap_fn` function shows up in the stack trace,
+            # if it's only once - means it's the top level mlflow call
+
+            wrap_fn_met = False
+            for stack_line in tb:
+                is_wrap_func = stack_line.filename == __file__ and stack_line.name in [
+                    "wrap_fn",
+                    "context_manager_wrap_fn",
+                ]
+                if not is_wrap_func:
+                    continue
+                elif wrap_fn_met:
+                    return False
+                else:
+                    wrap_fn_met = True
+
+            return True
+
+        def wrap_fn(*args, **kwargs):
+            try:
+                return fn(*args, **kwargs)
+            except MlflowException:
+                # Allow exceptions to bubble up if it's not the top-level user made call
+                if not is_top_level_mlflow_call():
+                    raise
+                logger.exception(
+                    f"MlflowException occurred when calling {fn.__name__} "
+                    f"but was suppressed due to running dagshub.mlflow.patch_mlflow:"
+                )
+
+        def context_manager_wrap_fn(*args, **kwargs):
+            # Context managers (start_run), always return a value, so those we retry for a long amount of time instead
+            # Swallowing this function's exception would break any call after that
+            # Alternative might be returning a mock object that no-ops any calls to it
+            for attempt in Retrying(stop=stop_after_delay(60 * 15), reraise=True):
+                with attempt:
+                    return fn(*args, **kwargs)
+
+        if fn.__name__ in self.context_manager_functions:
+            return context_manager_wrap_fn
+        else:
+            return wrap_fn
+
+    def is_function_already_patched(self, func_name: str) -> bool:
+        return func_name in self.original_func_lookup
+
+    def monkey_patch_module(self, module: ModuleType):
+        from mlflow.utils.lazy_load import LazyLoader
+
+        # Add module into set of patched modules, so there's no circular import recursion
+        self.patched_modules.add(module.__name__)
```

Comment:
1. You should probably use mutexes in these methods that check and modify the shared dict and set of patched methods and modules
2. What if it's already in the set? Shouldn't you skip the rest of this logic?

Reply:
1. This is not happening live at user-call-time, it's a one-off function at init level. Don't see a point in trying to make it thread-safe. Putting a lock on the patch/unpatch functions will have the same result.
2. It is being skipped, see line 125. Before I recurse into this function with a new module, I see that it wasn't patched already